### PR TITLE
Updating templates and sample apps for cart transform and discounts to use 2024-01

### PR DESCRIPTION
**Pull request by stefanlegg**
This PR updates the following to use the 2024-01 api version:

**Templates**

- checkout/*/cart-transform

**Sample apps**

- discounts
- bundles-cart-transform
- bundles-price-per-component
- optional-add-ons-cart-transform
- update-line-item

Additionally some minor changes were made to narrow type annotations more appropriately.

### CODE REVIEW

File: sample-apps/bundles-price-per-component/extensions/price-per-component-js/src/run.js [hunk 1]
```diff
@@ -44,10 +44,12 @@ export function run(input) {
...
 
+/**
+ * @param {RunInput['cart']['lines'][number]} cartLine
+ * @param {RunInput['presentmentCurrencyRate']} presentmentCurrencyRate
+ */
 function optionallyBuildExpandOperation({ id: cartLineId, merchandise }, presentmentCurrencyRate) {
-  const hasBundleDataMetafield = !!merchandise.product.bundledComponentData;
-
-  if (merchandise.__typename === "ProductVariant" && hasBundleDataMetafield) {
+  if (merchandise.__typename === "ProductVariant" && merchandise.product.bundledComponentData) {
```
Comment (stefanlegg) [inline sample-apps/bundles-price-per-component/extensions/price-per-component-js/src/run.js:52]: [note] adding better type annotations / narrowing

File: sample-apps/update-line-item/extensions/update-line-item-js/src/run.graphql [hunk 2]
```diff
@@ -3,7 +3,7 @@ query RunInput {
     lines {
       id
       cost {
-        totalAmount {
+        amountPerQuantity {
```
Comment (stefanlegg) [inline sample-apps/update-line-item/extensions/update-line-item-js/src/run.graphql:6]: [note] oops - we were perviously asking for the wrong thing

File: sample-apps/update-line-item/extensions/update-line-item-js/src/run.js [hunk 3]
```diff
@@ -46,6 +46,9 @@ export function run(input) {
   return operations.length > 0 ? { operations } : NO_CHANGES;
 };
 
+/**
+ * @param {RunInput['cart']['lines'][number]} cartLine
+ */
```
Comment (stefanlegg) [inline sample-apps/update-line-item/extensions/update-line-item-js/src/run.js:51]: [note] adding better type annotations

**Review by aaackerman — COMMENTED** [review]
Looks like a couple failures on the rust side for cart transform. Will wait to approve after those are fixed up!

**Review by aaackerman — APPROVED** [review]
Thanks for updating this! 🧹 🎉

**Review by zaquille-oneil — APPROVED** [review]

**Pull request merged**